apps/antalmanac/src/lib/enrollmentHistory.ts

```diff
@@ -0,0 +1,154 @@
+import { queryGraphQL } from './helpers';
+import { termData } from './termData';
+
+// This represents the enrollment history of a course section during one quarter
+export interface EnrollmentHistoryGraphQL {
+    year: string;
+    quarter: string;
+    department: string;
+    courseNumber: string;
+    dates: string[];
+    totalEnrolledHistory: string[];
+    maxCapacityHistory: string[];
+    waitlistHistory: string[];
+    instructors: string[];
+}
+
+export interface EnrollmentHistoryGraphQLResponse {
+    data: {
+        enrollmentHistory: EnrollmentHistoryGraphQL[];
+    };
+}
+
+// To organize the data and make it easier to graph the enrollment
+// data, we can merge the dates, totalEnrolledHistory, maxCapacityHistory,
+// and waitlistHistory arrays into one array that contains the enrollment data
+// for each day
+export interface EnrollmentHistory {
+    year: string;
+    quarter: string;
+    department: string;
+    courseNumber: string;
+    days: EnrollmentHistoryDay[];
+    instructors: string[];
+}
+
+export interface EnrollmentHistoryDay {
+    date: string;
+    totalEnrolled: number;
+    maxCapacity: number;
+    waitlist: number | null;
+}
+
+class _EnrollmentHistory {
+    // Each key in the cache will be the department and courseNumber concatenated
+    enrollmentHistoryCache: Record<string, EnrollmentHistory>;
+    termShortNames: string[];
+
+    constructor() {
+        this.enrollmentHistoryCache = {};
+        this.termShortNames = termData.map((term) => term.shortName);
+    }
+
+    queryEnrollmentHistory = async (department: string, courseNumber: string): Promise<EnrollmentHistory | null> => {
+        const cacheKey = department + courseNumber;
+        if (cacheKey in this.enrollmentHistoryCache) {
+            return this.enrollmentHistoryCache[cacheKey];
+        }
```

Comment:
A code density optimization might be to implement the class like this:
```suggestion
        return this.enrollmentHistoryCache[cacheKey] ??= this.findEnrollmentHistory(...)
```

You could delegate the search to a separate method, and apply nullish coalescing. The result of an assignment is the value that's assigned. If the cache entry exists, it'll be returned; otherwise the method will run to evaluate value, assign it to the cache, and get returned.